File: functions/src/game/logic.js
const RankValues = { 'two': 2, 'three': 3, 'four': 4, 'five': 5, 'six': 6, 'seven': 7, 'eight': 8, 'nine': 9, 'ten': 10, 'jack': 11, 'queen': 12, 'king': 13, 'ace': 14 };

function evaluateLiftWinner(plays, leadSuit, trumpSuit) {
  let winnerId = null;
  let bestCard = null;

  for (const [uid, card] of Object.entries(plays)) {
    if (!bestCard) {
      bestCard = card;
      winnerId = uid;
      continue;
    }

    const isTrump = card.suit === trumpSuit;
    const bestIsTrump = bestCard.suit === trumpSuit;

    if (isTrump && !bestIsTrump) {
      bestCard = card; winnerId = uid;
    } else if (isTrump && bestIsTrump) {
      if (RankValues[card.rank] > RankValues[bestCard.rank]) {
        bestCard = card; winnerId = uid;
      }
    } else if (!isTrump && !bestIsTrump) {
      if (card.suit === leadSuit && bestCard.suit !== leadSuit) {
        bestCard = card; winnerId = uid;
      } else if (card.suit === leadSuit && bestCard.suit === leadSuit) {
        if (RankValues[card.rank] > RankValues[bestCard.rank]) {
          bestCard = card; winnerId = uid;
        }
      }
    }
  }
  return winnerId;
}

module.exports = { evaluateLiftWinner };

File: functions/lib/game/logic.dart
import 'deck.dart';

const rankValues = <Rank, int>{
  Rank.two: 2,
  Rank.three: 3,
  Rank.four: 4,
  Rank.five: 5,
  Rank.six: 6,
  Rank.seven: 7,
  Rank.eight: 8,
  Rank.nine: 9,
  Rank.ten: 10,
  Rank.jack: 11,
  Rank.queen: 12,
  Rank.king: 13,
  Rank.ace: 14,
};

String? evaluateLiftWinner({
  required Map<String, Card> plays,
  required Suit leadSuit,
  required Suit trumpSuit,
}) {
  String? winnerId;
  Card? bestCard;

  for (final entry in plays.entries) {
    final uid = entry.key;
    final card = entry.value;

    if (bestCard == null) {
      bestCard = card;
      winnerId = uid;
      continue;
    }

    final isTrump = card.suit == trumpSuit;
    final bestIsTrump = bestCard.suit == trumpSuit;

    if (isTrump && !bestIsTrump) {
      bestCard = card;
      winnerId = uid;
    } else if (isTrump && bestIsTrump) {
      if (rankValues[card.rank]! > rankValues[bestCard.rank]!) {
        bestCard = card;
        winnerId = uid;
      }
    } else if (!isTrump && !bestIsTrump) {
      if (card.suit == leadSuit && bestCard.suit != leadSuit) {
        bestCard = card;
        winnerId = uid;
      } else if (card.suit == leadSuit && bestCard.suit == leadSuit) {
        if (rankValues[card.rank]! > rankValues[bestCard.rank]!) {
          bestCard = card;
          winnerId = uid;
        }
      }
    }
  }
  return winnerId;
}
